lib/TPP/Transforms/Vectorization.cpp

```diff
@@ -0,0 +1,109 @@
+
+namespace mlir {
+namespace tpp {
+
+struct LinalgGenericToVector : OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    if (xsmm::utils::getDataType(rewriter, linalgOp.getOperand(0).getType()) ==
+            xsmm::DataTypeAttr::get(rewriter.getContext(),
+                                    xsmm::DataType::BF16) &&
+        linalgOp.getIteratorTypes().size() >= 5) {
+      SmallVector<int64_t> shape;
+      SmallVector<ReassociationIndices> indices;
+      int index = 0;
+      for (size_t i = 0; i < dyn_cast<ShapedType>(linalgOp.getOperand(0).getType())
+                              .getShape()
+                              .size();
+           i++) {
+        ReassociationIndices reassoc;
+        if (i == dyn_cast<ShapedType>(linalgOp.getOperand(0).getType())
+                         .getShape()
+                         .size() -
+                     1) {
+          shape.push_back(dyn_cast<ShapedType>(linalgOp.getOperand(0).getType())
+                              .getShape()[i] /
+                          2);
+          shape.push_back(2);
+          reassoc.push_back(index++);
+          reassoc.push_back(index++);
+        } else {
+          shape.push_back(dyn_cast<ShapedType>(linalgOp.getOperand(0).getType())
+                              .getShape()[i]);
+          reassoc.push_back(index++);
+        }
+        indices.push_back(reassoc);
+      }
+      auto expand = rewriter.create<memref::ExpandShapeOp>(
+          linalgOp.getLoc(), shape, linalgOp.getOperand(0), indices);
+      linalgOp.setOperand(0, expand.getResult());
+      auto map0 = linalgOp.getIndexingMapsArray()[0];
+      auto map1 = linalgOp.getIndexingMapsArray()[1];
+      map0 = map0.insertResult(map1.getResult(map1.getNumResults() - 1), 3);
+      map1 = map1.insertResult(
+          dyn_cast<AffineBinaryOpExpr>(map1.getResult(1)).getLHS(), 2);
+      map1 = map1.dropResult(1);
+      linalgOp.setIndexingMapsAttr(rewriter.getAffineMapArrayAttr(
+          {map0, map1, linalgOp.getIndexingMapsArray()[2]}));
+    }
+    return linalg::vectorize(rewriter, linalgOp);
+  }
+};
+
+template <typename LinalgOp>
+struct LinalgToVector : OpRewritePattern<LinalgOp> {
+  using OpRewritePattern<LinalgOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(LinalgOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    return linalg::vectorize(rewriter, linalgOp);
+  }
+};
+
+struct VectorizationPass
+    : public impl::VectorizationPassBase<VectorizationPass> {
+
+  void populateCombinePatterns(RewritePatternSet &patterns) {
+    patterns.add<LinalgToVector<linalg::BatchReduceMatmulOp>,
+                 LinalgToVector<linalg::FillOp>>(patterns.getContext());
+    patterns.add<LinalgGenericToVector>(patterns.getContext());
```

Comment:
These patterns need more preconditions e.g., check for buffer semantics, currently there are hardcoded assumptions on the affine maps.

Reply:
Added bufferizability check, currently works with and without tiling pass.